src/commons/metrics-helper.ts

```diff
@@ -0,0 +1,61 @@
+import { StateDetails } from '../data-types/interfaces';
+import { logAndPublishMetric } from './metrics';
+import { EventsEnum, MetricNames, noMetadata } from '../data-types/constants';
+
+/**
+ * Function that compares the old account state and new account state and increments/decrements count of accounts blocked AND/OR accounts suspended accordingly
+ * @param oldState - the StateDetails representation of the user account's state before the applied event
+ * @param newState - the StateDetails representation of the user account's state after the applied event
+ */
+export function updateAccountStateCountMetric(oldState: StateDetails, newState: StateDetails) {
+  const blockedChange = (Number(oldState.blocked) - Number(newState.blocked)) * -1;
+  if (blockedChange !== 0) logAndPublishMetric(MetricNames.ACCOUNTS_BLOCKED, noMetadata, blockedChange);
+  const suspendChange = (Number(oldState.suspended) - Number(newState.suspended)) * -1;
+  if (suspendChange !== 0) logAndPublishMetric(MetricNames.ACCOUNTS_SUSPENDED, noMetadata, suspendChange);
+}
+
+/**
+ * Function that inspects the account information and decrements the appropriate count in response to the account being deleted
+ * @param blocked - boolean flag tracking if account is blocked
+ * @param suspended - boolean flag tracking if account is suspended
+ */
+export function updateAccountStateCountMetricAfterDeletion(blocked: boolean, suspended: boolean) {
+  if (blocked) logAndPublishMetric(MetricNames.ACCOUNTS_BLOCKED, noMetadata, -1);
+  else if (suspended) logAndPublishMetric(MetricNames.ACCOUNTS_SUSPENDED, noMetadata, -1);
+}
+
+/**
+ * Published the appropriate Time To Resolve metric when either an explicit suspension is removed and/or when a successful
+ * user led action has occurred.
+ * @param oldState - the previous account state object before the update
+ * @param currentState - the newly updated account state object
+ * @param oldAppliedAt - the timestamp when the previous account state was applied
+ * @param currentAppliedAt - the timestamp when the newly updated account state was applied at
+ * @param eventName - intervention name
+ */
+export function publishTimeToResolveMetrics(
+  oldState: StateDetails,
+  currentState: StateDetails,
+  oldAppliedAt: number,
+  currentAppliedAt: number,
+  eventName: string,
+) {
+  const timeDifferenceAsSeconds = Math.floor((currentAppliedAt - oldAppliedAt) / 1000);
+
+  if (oldState.suspended && !currentState.suspended) {
```

Comment:
hey Taher - where an user action event is causing the user to go from (for example) Suspended Psw Reset to Unsuspended, then we would be logging this metric twice, I think? Once with type `suspension` and once with type `resetPassword` - is this the intended behaviour?

Reply:
Hi @chiarapascucci-gds - yes - one metric to capture the time to resolve with a user led action and another to capture the time to unsuspend an account (either through user led action or when fraud remove suspension explicitly) - we may get only one metric published if the account is still suspended